core/lib/resources/presence/sentry.js

```diff
@@ -1,159 +1,270 @@
+//
+Sentry.prototype._loadAndCleanUpSentries = function(callback) {
   var self = this;
-  Persistence.readHashAll(Sentry.channel, function(replies) {
+
+  Persistence.readHashAll(redisSentriesKey, function(replies) {
     replies = replies || {};
-    Object.keys(replies).forEach(function(name) {
-      sentries[name] = replies[name];
+    var repliesKeys = Object.keys(replies);
 
-      // Cleanup
-      if (!self.isValid(name)) {
-        Persistence.deleteHash(Sentry.channel, name);
-        delete sentries[name];
+    self._purgeGoneSentries(replies, repliesKeys);
+
+    repliesKeys.forEach(function(name) {
+      self.sentries[name] = replies[name];
+      if (self.isDown(name)) {
+        self._purgeSentry(name);
       }
     });
 
-    if (callback) callback();
+    if (callback) { callback(); }
   });
 };
 
-// API
-Sentry.prototype.publishKeepAlive = function(options) {
-  options = options || {};
+Sentry.prototype._purgeSentry = function(name) {
+  var lastMessage;
+  
+  Persistence.deleteHash(redisSentriesKey, name);
 
-  // Optionally, publish on behalf of someone else
-  var name = options.name || this.name;
+  lastMessage = this.sentries[name];
+  logging.info('#presence - #sentry down:', name, lastMessage.host, lastMessage.port);
+  delete this.sentries[name];
+  this.emit('down', name, lastMessage);
+};
 
-  // Optionally, set custom expiry
-  var expiration = options.expiration || Date.now() + Sentry.expiry;
-  var host = this.host;
-  var port = this.port;
+// Deletion of a gone sentry key might just happened, so 
+// we compare existing sentry names to reply names
+// and clear whatever we have that no longer exists. 
+Sentry.prototype._purgeGoneSentries = function(replies, repliesKeys) {
+  var self = this,
+      sentriesGone = _.difference(this.sentryNames(), repliesKeys);
 
-  var message = {
-    name: name,
-    alive: true,
-    host: host,
-    port: port,
-    expiration: expiration
-  };
+  sentriesGone.forEach(function(name) {
+    self._purgeSentry(name);
+  });
+};
 
-  this.sentries[name] = message;
+// Listening for new pub sub messages from redis. 
+// As of now, we only care about new sentries going online. 
+// Everything else gets inferred based on time. 
+Sentry.prototype._startListening = function() {
+  var self = this;
 
-  // Specific check
-  if (options.save === false) return;
+  if (!this._listener) {
+    this._listener = function(channel, message) {
+      if (channel !== redisSentriesKey) {
+        return;
+      }
+      self._saveMessage(parseJSON(message));
+    };
 
-  Persistence.persistHash(Sentry.channel, this.name, message);
-  Persistence.publish(Sentry.channel, message);
-};
+    Persistence.pubsub().on('message', this._listener);
+  }
 
-Sentry.prototype.start = function(callback) {
-  var sentries = this.sentries;
-  var self = this;
-  if (this.timer) {
-    return;
+  if (!this._subscribed) {
+    this._subscribed = true;
+    Persistence.pubsub().subscribe(redisSentriesKey);
```

Comment:
you can push this into if(!this._listener) as well. No need for separate state (unless some of your tests need it..?)